app/src/main/java/org/p2p/wallet/swap/ui/jupiter/info/SwapInfoBottomSheet.kt

```diff
@@ -0,0 +1,210 @@
+        super.onViewCreated(view, savedInstanceState)
+        binding.recyclerViewRoutes.attachAdapter(adapter)
+        binding.buttonDone.setOnClickListener { dismiss() }
+        when (swapInfoType) {
+            SwapInfoType.MINIMUM_RECEIVED -> {
+                adapter.items = mapper.mapMinimumReceived()
+                binding.buttonDone.setText(R.string.swap_info_details_minimum_received_fee_done)
+            }
+            SwapInfoType.NETWORK_FEE -> {
+                adapter.items = mapper.mapNetworkFee()
+                binding.buttonDone.setText(R.string.swap_info_details_network_fee_done)
+            }
+            SwapInfoType.ACCOUNT_FEE -> {
+                adapter.items = mapper.mapAccountFee()
+                binding.buttonDone.setText(R.string.swap_info_details_account_fee_done)
+            }
+            SwapInfoType.LIQUIDITY_FEE -> {
+                binding.buttonDone.setText(R.string.swap_info_details_liquidity_fee_done)
+                observeFeatureState()
+            }
+        }
+    }
+
+    private fun observeFeatureState() {
+        val allTokens = flow { emit(interactor.getAllTokens()) }
+        stateManager.observe().combine(allTokens) { state, tokens ->
+            state to tokens
+        }
+            .flatMapLatest {
+                val (state, tokens) = it
+                handleFeatureState(state, tokens)
+            }
+            .flowOn(coroutineDispatchers.io)
+            .onEach { adapter.items = it }
+            .flowWithLifecycle(lifecycle)
+            .launchIn(lifecycleScope)
+    }
+
+    private suspend fun handleFeatureState(state: SwapState, tokens: List<SwapTokenModel>): Flow<List<AnyCellItem>> {
+        return when (state) {
+            SwapState.InitialLoading,
+            is SwapState.LoadingRoutes,
+            is SwapState.TokenAZero -> flowOf(mapper.mapEmptyLiquidityFee())
+            is SwapState.SwapException -> handleFeatureState(state.previousFeatureState, tokens)
+            is SwapState.SwapLoaded -> {
+                val route: JupiterSwapRoute = state.routes.getOrNull(state.activeRoute)
+                    ?: return flowOf(mapper.mapEmptyLiquidityFee())
+                flow {
+                    val rateLoaderList = mutableListOf<Flow<LoadRateBox>>()
+                    val feeCells = route.marketInfos.map { marketInfo ->
+                        val loadingCell = mapper.getLiquidityFeeCell(marketInfo, tokens)
+                        rateLoaderList += getRateLoaderFlow(marketInfo, tokens, loadingCell)
+                        loadingCell
+                    }
+                    val fullUiList = mapper.mapEmptyLiquidityFee().plus(feeCells)
+                    emit(fullUiList)
+
+                    rateLoaderList.merge()
+                        .collect {
+                            val marketInfo = it.first
+                            val loadingCell = it.second
+                            val rateLoaderState = it.third
```

Comment:
maybe put it like (marketInfo, loadingCell, rateLoaderState)?

Reply:
it doesn't work for Triple